File: Exercises/9/python_files/vehicles_polymorphic.py
class Vehicle:
    def __init__(self, wheel_size, tyre_thickness, name):
        self.wheel_size = wheel_size
        self.tyre_thickness = tyre_thickness
        self.name = name

    def show_details(self):
        print(f"Vehicle: {self.name}")
        print(f"Wheel Size: {self.wheel_size}")
        print(f"Tyre Thickness: {self.tyre_thickness}")


class Truck(Vehicle): # Truck INHERITS from Vehicle
    def __init__(self, wheel_size, tyre_thickness, name, cargo_capacity):
        # The following line calls the Vechicle constructor
        super().__init__(wheel_size, tyre_thickness, name)

		# Set the truck-specific attribute here
        self.cargo_capacity = cargo_capacity

    def show_details(self):
        super().show_details()
        print(f"Cargo Capacity: {self.cargo_capacity}kg")


class Car(Vehicle): # Car inherits from Vehicle
    def __init__(self, wheel_size, tyre_thickness, name, seating_capacity):
        # The following line calls the Vechicle constructor
        super().__init__(wheel_size, tyre_thickness, name)

		# Set the car-specific attribute here
        self.seating_capacity = seating_capacity

    def show_details(self):
        super().show_details()
        print(f"Seating Capacity: {self.seating_capacity}")

# Testing code
def display_vehicles(vehicles):
    for vehicle in vehicles:
        vehicle.show_details()          # this line is polymorphic
        print()

truck = Truck(34, 22, "Izuzu", 3435.0)
car = Car(80, 30, "Toyota", 5)
vehicles = [truck, car]

display_vehicles(vehicles)

File: Exercises/9/JavaFiles/VehiclesPolymorphic.java
package JavaFiles;
import java.util.ArrayList;


class Vehicle {
    protected int wheelSize;
    private int tyreThickness;
    private String name;

    public Vehicle(int wheelSize, int tyreThickness, String name) {
        this.wheelSize = wheelSize;
        this.tyreThickness = tyreThickness;
        this.name = name;
    }

    public void showDetails() {
        System.out.println("Vehicle: " + name);
        System.out.println("Wheel Size: " + wheelSize);
        System.out.println("Tyre Thickness: " + tyreThickness);
    }
}


class Truck extends Vehicle {
    private int cargoCapacity;

    public Truck(int wheelSize, int tyreThickness, String name, int cargoCapacity) {
        super(wheelSize, tyreThickness, name); // Call the constructor of the Vehicle class
        this.cargoCapacity = cargoCapacity;
    }

    @Override
    public void showDetails() {
        super.showDetails();
        System.out.println("Cargo Capacity: " + cargoCapacity + "kg");
    }
}


class Car extends Vehicle {
    private int seatingCapacity;

    public Car(int wheelSize, int tyreThickness, String name, int seatingCapacity) {
        super(wheelSize, tyreThickness, name); // Call the constructor of the Vehicle class
        this.seatingCapacity = seatingCapacity;
    }

    @Override
    public void showDetails() {
        super.showDetails();
        System.out.println("Seating Capacity: " + seatingCapacity);
    }
}

class Garage {
    private ArrayList<Vehicle> vehicles;

    Garage() {
        vehicles = new ArrayList<>();
    }

    void addVehicle(Vehicle vehicle) {
        this.vehicles.add(vehicle);
    }

    void displayVehicles() {
        for (Vehicle vehicle : vehicles) {
            vehicle.showDetails();
            System.out.println(); // Just to give some separation between outputs
        }
    }
}

public class VehiclesPolymorphic
{
    public static void main(String[] args) {
        Garage garage = new Garage();
        Truck truck = new Truck(34, 22, "Izuzu", 3435);
        Car car = new Car(80, 30, "Toyota", 5);

        garage.addVehicle(truck);
        garage.addVehicle(car);

        garage.displayVehicles();
    }
}